packages/nextcloud/lib/src/helpers/spreed.dart

```diff
@@ -174,30 +194,70 @@ enum ParticipantPermission {
 
   /// Integer representation of the [ParticipantPermission].
-  int get value => index == 0 ? 0 : 1 << (index - 1);
+  int get value => _bitValueFromEnumIndex(index);
 
   /// Converts the integer representation of multiple [ParticipantPermission]s to the corresponding [ParticipantPermission]s.
-  static Set<ParticipantPermission> fromValue(final int value) {
-    final permissions = <ParticipantPermission>{};
-
-    var v = value;
-    for (var i = 1; i <= ParticipantPermission.values.length - 1; i++) {
-      if (v.isOdd) {
-        permissions.add(ParticipantPermission.values[i]);
-      }
-      v = v >> 1;
-    }
-
-    if (permissions.isEmpty) {
-      permissions.add(ParticipantPermission.$default);
-    }
-
-    return permissions;
-  }
+  static Set<ParticipantPermission> fromValue(final int value) => _enumFromBitValue(values, value);
 }
 
 /// Extension for the integer representation of multiple [ParticipantPermission]s.
 extension ParticipantPermissionsValue on Set<ParticipantPermission> {
   /// Gets the integer representation of multiple [ParticipantPermission]s.
   int get value => map((final p) => p.value).reduce((final a, final b) => a | b);
 }
+
+/// Participant in-call flags.
+///
+/// Use [fromValue] to convert the integer representation into this enum representation.
+/// Use [value] to get the integer representation that is used in the API.
+/// Use [ParticipantInCallFlagsValue.value] to convert multiple [ParticipantInCallFlag] into the integer representation.
+///
+/// See https://github.com/nextcloud/spreed/blob/main/lib/Participant.php.
+enum ParticipantInCallFlag {
+  /// Not connected to the call.
+  disconnected,
+
+  /// Connected to the call.
+  inCall,
+
+  /// Connected to the call with audio.
+  providesAudio,
+
+  /// Connected to the call with video.
+  providesVideo,
+
+  /// Connected to the call using SIP dial-in.
+  sipDialIn;
+
+  /// Integer representation of the [ParticipantInCallFlag].
+  int get value => _bitValueFromEnumIndex(index);
+
+  /// Converts the integer representation of multiple [ParticipantInCallFlag]s to the corresponding [ParticipantInCallFlag]s.
+  static Set<ParticipantInCallFlag> fromValue(final int value) => _enumFromBitValue(values, value);
+}
+
+/// Extension for the integer representation of multiple [ParticipantInCallFlag]s.
+extension ParticipantInCallFlagsValue on Set<ParticipantInCallFlag> {
+  /// Gets the integer representation of multiple [ParticipantInCallFlag]s.
+  int get value => map((final p) => p.value).reduce((final a, final b) => a | b);
+}
+
+int _bitValueFromEnumIndex(final int index) => index == 0 ? 0 : 1 << (index - 1);
+
+Set<T> _enumFromBitValue<T extends Enum>(final List<T> values, final int value) {
+  final result = <T>{};
+
+  var v = value;
+  for (var i = 1; i <= values.length - 1; i++) {
+    if (v.isOdd) {
+      result.add(values[i]);
+    }
+    v = v >> 1;
+  }
+
+  if (result.isEmpty) {
+    result.add(values.first);
+  }
+
+  return result;
+}
```

Comment:
Nice to generalize this but why not make it general extensions for all Enums?
This way we can reuse it anywhere.

the fromBinary must be an extension on List<T> (`EnumFromBinary<T extends Enum> on List<T>`) to get access to the values. This way we also wouldn't need to pass the values around.

This could be something like (caution untested):
```dart
extension EnumBinaryCollection<T extends Enum> on Set<T> {
  int get binaryValue => map((final p) => p.binaryValueFromEnumIndex).reduce((final a, final b) => a | b);
}

extension EnumToBinary on Enum {
  int get binaryValueFromEnumIndex => index == 0 ? 0 : 1 << (index - 1);
}

extension EnumFromBinary<T extends Enum> on List<T> {
  Set<T> enumFromBitValue(final int value) {
    final result = <T>{};

    var v = value;
    for (var i = 1; i <= length - 1; i++) {
      if (v.isOdd) {
        result.add(this[i]);
      }
      v = v >> 1;
    }

    if (result.isEmpty) {
      result.add(first);
    }

    return result;
  }
}
```

We do not need to export these helper methods so we don't confuse users to much.
The helper methods with `fromValue` and `value` inside each enum already handle that.

Reply:
By not exporting the extensions from the nextcloud package (putting them under `packages/nextcloud/lib/src/helpers/enum_helpers.dart` and annotate them with internal) We would be able to reuse them within the nc package but not need to think about any documentation.

The Enum needing them like `RoomType` would still implement the `fromValue` and to `value` methods that internally call these unexposed ones.

Does that make sense?

```dart
enum SomeBinaryEnum{
	value1,
	value2;

	/// Integer representation of the [SomeBinaryEnum].
	int get value => binaryValueFromEnumIndex;

	/// Converts the integer representation of multiple [SomeBinaryEnum]s to the corresponding [SomeBinaryEnum]s.
	static Set<SomeBinaryEnum> fromValue(final int value) => SomeBinaryEnum.values.enumFromBitValue(value);

}

@internal
extension EnumBinaryCollection<T extends Enum> on Set<T> {
  int get binaryValue => map((final p) => p.binaryValueFromEnumIndex).reduce((final a, final b) => a | b);
}

@internal
extension EnumToBinary on Enum {
  int get binaryValueFromEnumIndex => index == 0 ? 0 : 1 << (index - 1);
}

@internal
extension EnumFromBinary<T extends Enum> on List<T> {
  Set<T> enumFromBitValue(final int value) {
    final result = <T>{};

    var v = value;
    for (var i = 1; i <= length - 1; i++) {
      if (v.isOdd) {
        result.add(this[i]);
      }
      v = v >> 1;
    }
```

Reply:
This way we would have one internal method handling all of this (especially as the extensions `ParticipantPermissionsValue` and `ParticipantInCallFlagsValue` are already duplicating logic).

Reply:
I opted for making the extension public and documented. I added the information about the binary representation and how to handle them to each of the relevant enums.